lib/pbench/server/api/resources/users_api.py

```diff
@@ -440,15 +436,31 @@ def put(self, username):
                 }
         """
         post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
         try:
             user, verified = self.auth.verify_user(username)
         except Exception:
             self.logger.exception("Exception occurred while verifying the user")
             abort(500, message="INTERNAL ERROR")
 
         # TODO: Check if the user has the right privileges
-        if verified or user.is_admin():
+        if verified:
             try:
+                # Log if the user payload contain fields that are either non-updatabale or
+                # are not present in the user db.
+                non_updatable = list(
+                    set(post_data.keys()) - set(User.__table__.columns.keys())
+                )
+                if "registered_on" in post_data.keys():
+                    non_updatable.append("registered_on")
```

Comment:
Again, this means a client can't do

```
json = GET /user/id
json.password = "new"
PUT /user/id {json}
```

(in fairly loose notation...) The point is, this will fail because "they've specified" the non-updateable `registered_on` property. But a better way to handle this would be to fail only if it's specified __and__ the value differs from the current value in the DB.

This style requires creating a new JSON document with only the fields to be changed; in many cases (like the password reset we've been discussing) that style is going to be easier anyway, and for a relatively small object with little data that can be changed, it probably isn't a big deal. (I'm not sure we want to allow changing username or email, for example: changing username would mean editing every Elasticsearch document owned by the user, which doesn't seem worth supporting, and changing the email, given that we're using it for password reset, ought to mean validating the new email... maybe worth doing, but we can put that off. Actually, we should probably validate the email on registration, too; another thing to look at along with the password reset, because it'll also mean generating and sending an email with a link back to the server.)

Reply:
@dbutenhof it wont fail, the code would just skip the `registered_on` field and update the other fields. In the case of following 
```
json = GET /user/id
json.password = "new"
PUT /user/id {json}
```
the only real update will happen on the `password`. The `registered_on` key will be skipped and other fields such as  `first_name`, `last_name` etc will get updated to its current value (essentially meaning no update) the code doesn't check if there a difference between an old value and a new value, it just updates to whatever is specified. However, it wont fail.

Reply:
> I am hoping we are not going to use `username` field values in the elasticsearch owner data, but instead an immutable ID from the database.

I had considered that. I'm not, mostly because "users" don't actually exist and there is no ID. On the other hand, if `username` isn't mutable, it's a bit cleaner not to reference an internal non-reproducible ID value across separate databases. If we want someone to be able to change their username at will (or if we end up changing to the current dashboard user schema and using email as the username without a separate field at all), then I'd agree I'll need to change the indexer and queries when I merge with Nikhil's work.

Reply:
> @dbutenhof it wont fail, the code would just skip the `registered_on` field and update the other fields.

Yes; what I'm arguing is that attempting a `PUT` that changes a "protected" field _should_ be an error, while a `PUT` that includes a protected field with an unchanged value doesn't even merit a log, much less a warning.

Reply:
@dbutenhof you mean we return an error if a user is attempting to change the "protected" field and refuse to change other fields as well if the "protected" field is inside the json payload whose value is not same as the original value?

Reply:
> I am hoping we are not going to use username field values in the elasticsearch owner data, but instead an immutable ID from the database.

IMNSHO, username, user ID, and email address should all be separate things.  We can argue whether we should support the username being mutable (although, I would be modestly in favor), but the email address almost certainly should be mutable, and using the email address as the user ID can result in a conflict (or, at least, a strange juxtaposition) when the user needs to change his email address (as we do, from time to time).

So, I would lobby for identifying users with an immutable user ID, and then hanging the username and email address off of that.  Yes, it's a good bit more work, especially if we're maintaining multiple databases, but, in the end, I think it will be worth it.

Reply:
> @dbutenhof you mean we return an error if a user is attempting to change the "protected" field and refuse to update other fields as well if the "protected" field is inside the json payload whose value is not same as the original value?

Yes; if a `PUT` payload attempts to change a protected field, that's an error; we shouldn't change anything.

Ideally, at some point we'll consider generating e-Tag values that will allow protection against asynchronous PUTs so we can fail any `PUT` with obsolete data (that would overwrite a recent change either internally or externally); I don't want to think about that right now, though. ;-)

Reply:
> So, I would lobby for identifying users with an immutable user ID, and then hanging the username and email address off of that. Yes, it's a good bit more work, especially if we're maintaining multiple databases, but, in the end, I think it will be worth it.

We have the Postgres-generated row `id` value, which we could use as-is elsewhere; or we could generate some unique string value that might be easier to read. Arguably, we could go "low tech" and simply duplicate the client-specified `username` field as `userid` and `username` with the former held immutable as a permanent internal key and the latter as a mutable "display" value that can be freely changed without affecting the internal references.

Reply:
Any of those would fit the bill, I think; however, if you use the initial `username` as the `userid`, then you need some way to resolve collisions when a new user tries to create a login using a `username` value which matches some other user's `userid`.  Presumably we want all currently-existing `username` values to be unique (for UX reasons), but does a `username` have to be unique across all historical ones?...maybe...(e.g., to prevent squatting and spoofing).

Reply:
> Any of those would fit the bill, I think; however, if you use the initial `username` as the `userid`, then you need some way to resolve collisions when a new user tries to create a login using a `username` value which matches some other user's `userid`. Presumably we want all currently-existing `username` values to be unique (for UX reasons), but does a `username` have to be unique across all historical ones?...maybe...(e.g., to prevent squatting and spoofing).

Ha; I actually deleted an extra sentence or two about that, which I ultimately decided might be excessive verbiage at the time.

Yes; if I create "dave" and then change to "david", we've got two "unique" values to check and maintain. I think we'd treat `userid` as an internal value; whether we actually hide it on a `GET` is a different matter, but, like if we'd used the integer Postgres row `id`, we don't need to worry about how it looks. So my original "dave" account when edited would have `"userid": "dave", "username": "david"` and when a new "dave" is created it could be transparently uniqueified to `"userid": "dave1", "username": "dave"`.

If we don't persist "username" outside the User DB model, I don't think we need to worry about historical values; however to simplify bookkeeping ... do we need to worry about historical `userid` values, even if someone deletes a User? Hmm. I'm not really sure what we do with Elasticsearch documents if someone deletes a User. Do we immediately erase all "private" documents owned by that user? Do we explicitly "orphan" them?

But yeah, if we _don't_ modify the ownership of those documents, then when someone creates a new account with a repeat `userid` they implicitly inherit all those documents, which would be bad. Does that mean we just mark a "deleted" User as `deleted` without actually removing it from the DB? Even if we use the row `id`, will Postgres/SQLAlchemy ever re-use a row id when rows are deleted? I don't really like the idea of permanently retaining deleted User records just for that, but maybe we need to, whatever type of userid we choose, to avoid this.

Reply:
Yes, coping with _deleted_ users is a whole 'nother ball of wax.  However, what I was thinking about is the case where a user changes his/her username and then another user appears who (re)uses the previous user's old username:  to the extent that one user is aware of other users (think social media platform...), then having a user change his/her username make him/her hard to follow/track, but then having a new user show up with the old username creates confusion because other users may not notice the change of identity.

This might not be a problem for us.  To the extent that the username is just "presentation sugar" which improves the UX for the single user, then we probably don't care about user spoofing.  But, if we're going to offer the ability to share data between users, then we need to enable users to be sure of the identity of the people with whom they are intending to share.  If we offer the ability for a user to share data, then the user has to be able to pick the people, and presumably that would _not_ be done via `userid` but would instead be done by `username`...but, if the `username` value is mutable and not historically-unique, then I'm not sure how a user can be sure that the target `username` is the person that s/he thinks it is.

As for deleting users, I think it would be best to retain the deleted users and mark them as deleted.  I don't think that this will be a heavy burden, and it prevents problems with reused `userid`'s and possibly `username`'s.  What to do with their data is a trickier question, and whether it has been shared will feed into the answer.

Reply:
> [...] If we offer the ability for a user to share data, then the user has to be able to pick the people, and presumably that would _not_ be done via `userid` but would instead be done by `username`...but, if the `username` value is mutable and not historically-unique, then I'm not sure how a user can be sure that the target `username` is the person that s/he thinks it is.

Ah; good point. While it'd be easy to expect that Pbench servers will always be a relatively closed audience, that may not be true if we get pulled into the greater Red Hat cloud provider ecosystem. Although I have no idea how we'll need to manage identities under that model, and it'd probably be with an OpenStack cloud identity manager rather than maintaining our own user database anyway.

But ... for now ... hmm. We could add a table of "claimed usernames" independent of any current User username, and simply never delete from that pool; or at least expire from that list only after some time has passed. Though that would mean that if I create an account, mess something up, and immediately delete it, I can't (ever?) create another account with the same username, which could be a pain. I don't think we want to be in the "one strike and you're out" territories.

As for sharing/spoofing ... right now every dataset is either "private" and locked to a specific owning `userid`, or it's "public" and accessible to everyone, so that's not an issue. If we allow sharing to specific users (or hypothetical groups), I suspect we can solve the problem of "inherited sharing" by actually sharing to the permanently unique `userid` even though the UI might specify a currently active `username`. If someone else later takes that username, they won't get previously shared data.

And yeah, if you delete your "webb" user and I later create a new "webb" (with `userid` "webb1"), someone might still accidentally share with me thinking I'm you... we can solve that by the "claimed username" history (which, as I said, might also prevent you from reclaiming the "webb" User), or we can simply leave that issue to community/admin policing because I think any solution in the code might be unrealistically limiting.

Reply:
I think the take-away here is that users should be identified internally (for authorization, access control, data ownership, etc.) by their `userid`, however that might be created and managed (it could be via an external identity manager, but initially it will be something which we produce, and it could be as straightforward as the Postgres row ID); however, end users will never deal with those values (directly), and instead will deal with `username` values, which we will translate (somehow, somewhere) to `userid`'s.  If we allow `username`'s to be mutable, then that brings in identity problems for second-parties, in that, once changed, the translation to `userid` may not be unique considering past history.

> if I create an account, mess something up, and immediately delete it, I can't (ever?) create another account with the same username, which could be a pain.

Yes, but it would not be without precedent.  On the other hand, it might be within the realm of possibility to support "undeleting" an account (I don't know how helpful this would be, depending on why the account was initially deleted).  If you can reset the password, I don't see any reason why you should not be able to reclaim the entire account (minus any data, of course, which would, presumably, have been disposed of when the account was deleted).

> If someone else later takes that username, they won't get previously shared data.

Right, but, as you explained they might get newly-shared data.  At this point, I'm thinking we should enable but not actually provide mutable `username` values.  If people cannot change them, then a bunch of UX and implementation problems never arise.  If only the UI uses `username`'s and if the server only uses `userid`'s (other than offering the service of translating between them), then I think we get everything we need along with the possibility of having things that we might want in the future, without creating additional problems in the present.

Reply:
> At this point, I'm thinking we should enable but not actually provide mutable `username` values. If people cannot change them, then a bunch of UX and implementation problems never arise. If only the UI uses `username`'s and if the server only uses `userid`'s (other than offering the service of translating between them), then I think we get everything we need along with the possibility of having things that we might want in the future, without creating additional problems in the present.

Agreed. And in terms of Nikhil's work, the immediate implication (for this PR or better a close follow-on) is to add the `userid` field, which is constrained to be a duplicate of `username` and which cannot be changed. When I merge, I'll use that to identify the ownership of Elasticsearch documents, and that indexing and query code won't be affected by any later scheme to allow changing `username`.

Reply:
I think we need to add grace period (something like 7 days or n number of days) to all the documents owned by a user especially private documents so when someone deletes a user the countdown will start on all the documents owned by that user and they will be deleted after the grace period is ended. So if a user decides to register back within a grace period can get the ownership back. However, we shouldn't orphan them as they never meant to be public, I guess many time when users delete themselves, they understand that any documents attached to the account will also be deleted.

Reply:
I would hesitate to provide a grace period, for a number of reasons.  Aside from requiring an agency which actually determines that the countdown is complete and performs the deletion (without fail...), there are potential issues with the GDPR and/or similar regulations.  I think, if a user requests to have their account deleted,all unshared data should disappear along with it.  (We can discuss what should happen to shared data, but, since we don't have any of that, yet, it's not pressing.)  If there is a concern, we can make it very hard to delete an account "by accident" (and, we can issue lots of warnings during the process), but, once the account is gone, anything that it owns should be gone, too.

Reply:
Yes I agree and I think that's reasonable

Reply:
Resolving the conversation here we can continue to brain storm on this in the github discussions